api/src/server.rs

```diff
@@ -399,27 +407,45 @@ where
             // rely on the channel depth for backpressure. the goal is to keep the queue close to empty
             // without processing one at a time. when we stop parsing the carfile in the store
             // i.e. validate before sending here and this is just an insert, we may want to process more at once.
+            let mut shutdown = false;
             loop {
                 let mut buf = Vec::with_capacity(EVENTS_TO_RECEIVE);
+                let mut process_early = false;
                 tokio::select! {
                     _ = interval.tick() => {
-                        Self::process_events(&mut events, &event_store, node_id).await;
+                        process_early = true;
                     }
                     val = event_rx.recv_many(&mut buf, EVENTS_TO_RECEIVE) => {
                         if val > 0 {
                             events.extend(buf);
                         }
                     }
-                }
-                let shutdown = event_rx.is_closed();
-                // make sure the events queue doesn't get too deep when we're under heavy load
-                if events.len() >= EVENT_INSERT_QUEUE_SIZE || shutdown {
-                    Self::process_events(&mut events, &event_store, node_id).await;
-                }
+                    _ = shutdown_signal.recv() => {
+                        tracing::debug!("Insert many task got shutdown signal");
+                        shutdown = true;
+                    }
+                };
                 if shutdown {
-                    tracing::info!("Shutting down insert task.");
+                    tracing::info!("Shutting down insert task after processing current batch");
+                    if !event_rx.is_empty() {
+                        let remaining_event_cnt = event_rx.len();
+                        let mut buf = Vec::with_capacity(remaining_event_cnt);
+                        tracing::info!(
+                            "Receiving {remaining_event_cnt} remaining events for insert task before exiting"
+                        );
+                        event_rx.recv_many(&mut buf, event_rx.len()).await;
+                        events.extend(buf);
+                    }
+
+                    Self::process_events(&mut events, &event_store, node_id).await;
                     return;
                 }
+                // process events at the interval or when we're under heavy load.
+                // we do it outside the select! to avoid any cancel safety issues
+                // even though we should be okay since we're using tokio channels/intervals
+                if events.len() >= EVENT_INSERT_QUEUE_SIZE || process_early {
```

Comment:
what's "early" about the processing when `process_early` is true?  Isn't the "early" processing really when the events.len() is too long, but the processing when that bool is true is actually the regularly scheduled processing due to the interval ticking?

Reply:
Fair. Process interval is a better name. I’ll rename it. It was early in my head because we want to get N requests before we do work so if we don’t get enough in time we process them anyway before that (ie early) but it’s not very clear.